cpp/src/arrow/compute/exec/key_compare.cc

```diff
@@ -17,250 +17,398 @@
+template <bool use_selection, bool is_first_varbinary_col>
+void KeyCompare::CompareVarBinaryColumnToRow(
+    uint32_t id_varbinary_col, uint32_t num_rows_to_compare,
+    const uint16_t* sel_left_maybe_null, const uint32_t* left_to_right_map,
+    KeyEncoder::KeyEncoderContext* ctx, const KeyEncoder::KeyColumnArray& col,
+    const KeyEncoder::KeyRowArray& rows, uint8_t* match_bytevector) {
+#if defined(ARROW_HAVE_AVX2)
+  if (ctx->has_avx2()) {
+    CompareVarBinaryColumnToRow_avx2(
+        use_selection, is_first_varbinary_col, id_varbinary_col, num_rows_to_compare,
+        sel_left_maybe_null, left_to_right_map, ctx, col, rows, match_bytevector);
+    return;
+  }
+#endif
 
-  for (uint32_t id_input = num_rows_already_processed; id_input < num_rows; ++id_input) {
-    uint32_t irow_left = use_selection ? sel_left_maybe_null[id_input] : id_input;
+  const uint32_t* offsets_left = col.offsets();
+  const uint32_t* offsets_right = rows.offsets();
+  const uint8_t* rows_left = col.data(2);
+  const uint8_t* rows_right = rows.data(2);
+  for (uint32_t i = 0; i < num_rows_to_compare; ++i) {
+    uint32_t irow_left = use_selection ? sel_left_maybe_null[i] : i;
     uint32_t irow_right = left_to_right_map[irow_left];
-    uint32_t begin_left = length * irow_left;
-    uint32_t begin_right = length * irow_right;
+    uint32_t begin_left = offsets_left[irow_left];
+    uint32_t length_left = offsets_left[irow_left + 1] - begin_left;
+    uint32_t begin_right = offsets_right[irow_right];
+    uint32_t length_right;
+    uint32_t offset_within_row;
+    if (!is_first_varbinary_col) {
+      rows.metadata().nth_varbinary_offset_and_length(
+          rows_right + begin_right, id_varbinary_col, &offset_within_row, &length_right);
+    } else {
+      rows.metadata().first_varbinary_offset_and_length(
+          rows_right + begin_right, &offset_within_row, &length_right);
+    }
+    begin_right += offset_within_row;
+    uint32_t length = std::min(length_left, length_right);
     const uint64_t* key_left_ptr =
         reinterpret_cast<const uint64_t*>(rows_left + begin_left);
     const uint64_t* key_right_ptr =
         reinterpret_cast<const uint64_t*>(rows_right + begin_right);
-    uint64_t result_or = 0ULL;
-    int32_t istripe = 0;
-
-    // Specializations for keys up to 8 bytes and between 9 and 16 bytes to
-    // avoid internal loop over words in the value for short ones.
-    //
-    // Template argument 0 means arbitrarily many 64-bit words,
-    // 1 means up to 1 and 2 means up to 2.
-    //
-    if (num_64bit_words == 0) {
-      for (; istripe < num_loops_less_one; ++istripe) {
-        uint64_t key_left = util::SafeLoad(&key_left_ptr[istripe]);
-        uint64_t key_right = util::SafeLoad(&key_right_ptr[istripe]);
-        result_or |= (key_left ^ key_right);
+    uint64_t result_or = 0;
+    if (length > 0) {
+      int32_t j;
+      // length can be zero
+      for (j = 0; j < (static_cast<int32_t>(length) + 7) / 8 - 1; ++j) {
```

Comment:
Nit: please replace all of these with calls to descriptively named helpers, like:
```suggestion
      for (j = 0; j < BitUtil::CeilDiv(length, 8); ++j) {
```

Reply:
done